### ISSUE TITLE
Buggy /back could be worth looking into

### ISSUE THREAD

**Issue by aschmois** [ForgeEssentials/ForgeEssentialsMain#1544]
Many times with me and a few friends the /back command will teleport you to ground level if you were in a slightly too enclosed space underground. It may or may not be a mod acting weird on this but I can't really reproduce it a lot, just every once in a while. Sounds like a good idea to get rid of these tiny bugs for beta. I'll take a look at the code while I'm at it, see if I can find anything.

**Comment by olee** [comment]
What do you mean to ground level?
Y == 0 or just a bit into the ground (i.e. one or two blocks lower than it's supposed to teleport)?

**Comment by luacs1998** [comment]
@aschmois can you take a look at this?

**Comment by aschmois** [comment]
I mean that when inside a cave I die, I teleport using /back and I appear outside of the cave with a view of the sky. Ground level means over the cave, sometimes 50 blocks over.

**Comment by olee** [comment]
Maybe the grave blocks the location to teleport back to, making the teleport handler trying to find a new location...

**Comment by aschmois** [comment]
A little bit of research, I would fix these things but I don't want to break anything else.

When a player dies these two methods get called
````
pi.setLastDeathLocation(new WarpPoint(player));
pi.setLastTeleportOrigin(pi.getLastDeathLocation());
````
We should remove setLastTeleportOrigin since /back already checks for last death location. It wouldn't change behavior but new behavior could break assuming that last teleport origin is different than last death location.

Also, by removing setLastTeleportOrigin the permission teleport back on death actually gets used.
````
        if (PermissionsManager.checkPermission(sender, TeleportModule.PERM_BACK_ONDEATH))
            point = pi.getLastDeathLocation();
        if (point == null)
            point = pi.getLastTeleportOrigin();
````
As you can see by setting last teleport origin, you are completely ignoring the first case here and the permission becomes useless.

Now for the actual teleportation.

When the player dies, a warp point gets created that saves the player location. There's absolutely nothing here that finds any safe location. Throughout the entire teleportation process, the warp point is never checked to see if teleporting the player to that place will actually place them in an unsafe position or not. There is no findSafeY (from today, it might have been removed). If the location of /back is blocking the teleport, the method would just say that the teleport is blocked and it would not actually teleport you, but it would just send a message to the user saying that the location is blocked.

So in conclusion, FE does things correctly and there is something else acting on this. When teleported, FE does not change the teleportation location one bit, but something else may be moving the player on death to ground level with a higher priority so when saving the player's death location, FE is saving the incorrect location. Do you have any other evens for player death than this one? If not, there has to be a mod acting on it.

**Comment by olee** [comment]
You are right - that stuff is really not done properly that.
The problem however why I set death- and teleport origin in the teleport function is, because we cannot do it in `/back`, because the teleport might be canceled during warmup.
I will check that whole stuff later.

**Comment by luacs1998** [comment]
Is this still a thing?

**Comment by olee** [comment]
I don't think so... I did change some stuff there some time ago so this might already be a no issue.
